app/pages/settings/email.jsx

```diff
@@ -0,0 +1,299 @@
+      }
+    })
+    .catch((error) => {
+      console.log('Something went wrong. Error: ', error);
+    });
+  }
+
+  handlePreferenceChange(preference, event) {
+    if (event.target.type === 'radio') {
+      const newTalkPrefs = Object.assign({}, this.state.talkPreferences);
+      preference.update({
+        email_digest: event.target.value
+      })
+      .save();
+      this.setState();
+    } else {
+      preference.update({
+        email_communication: !!event.target.value
+      })
+      .save();
+      this.setState();
+    }
+  }
+
+  nameOfPreference(preference) {
+    switch (preference.category) {
+      case 'participating_discussions':
+        return 'When discussions I\'m participating in are updated';
+      case 'followed_discussions':
+        return 'When discussions I\'m following are updated';
+      case 'mentions':
+        return 'When I\'m mentioned';
+      case 'group_mentions':
+        return 'When I\'m mentioned by group (this.admins, this.team, etc)';
+      case 'messages':
+        return 'When I receive a private message';
+      case 'started_discussions':
+        return 'When a discussion is started in a board I\'m following';
+      default:
+        return ' ';
+
+    }
+  }
+
+  renderProjectPreferences() {
+    const { projects, projectPreferences } = this.state;
+    return projectPreferences.map((projectPreference, i) => {
+      if (projects[i]) {
+        return (
+          <tr key={projectPreference.id}>
+            <td>
+              <input
+                type="checkbox"
+                name="email_communication"
+                checked={projectPreference.email_communication}
+                onChange={this.handlePreferenceChange.bind(this, projectPreference)}
+              />
+            </td>
+            <td>
+              {projects[i].display_name}
+            </td>
+          </tr>
+        );
+      }
+    });
+  }
+
+  renderPagination() {
+    const { meta, page } = this.state;
+    if (meta) {
+      return (
+        <nav className="pagination">{'Page'}
+          <select
+            value={page}
+            disabled={meta.page_count < 2}
+            onChange={e => this.setState({ page: e.target.value })}
+          >
+            {Array.from(Array(meta.page_count), (p, i) => {
+              return <option key={i} value={i + 1}>{i + 1}</option>;
+            })}
+          </select> {' of '} {meta.page_count || '?'}
+        </nav>
+      );
+    }
+  }
+
+  renderTalkPreferences() {
+    const { talkPreferences } = this.state;
+    const sortedPrefs = this.sortPreferences(talkPreferences);
+    return (
+      (sortedPrefs.length > 0) ?
+        <tbody>
+          {sortedPrefs.map((pref) => {
+            if (pref.category !== 'system' && pref.category !== 'moderation_reports') {
+              return (
+                <tr key={pref.id}>
+                  <td>{this.nameOfPreference(pref)}</td>
+                  {this.talkPreferenceOption(pref, 'immediate')}
```

Comment:
Each of these options will need a unique key.

Reply:
Also, as of React 14, functions that return components, like this one, can also be written as React components.

Reply:
Ignore me. The key warning is bad test data in the tests. I've fixed it on the `email-settings-tests` branch.